apiserver/common/crossmodel/auth.go

```diff
@@ -373,42 +404,86 @@ func (a *authenticator) checkMacaroons(
 	requiredRelation := requiredValues[relationKey]
 	requiredOffer := requiredValues[offeruuidKey]
 	requiredSourceModelUUID := requiredValues[sourcemodelKey]
-	authYaml, err := a.ctxt.offerPermissionYaml(requiredSourceModelUUID, username, requiredOffer, requiredRelation, permission.ConsumeAccess)
+
+	var m *bakery.Macaroon
+	if a.ctxt.jaasOfferBakery == nil {
+		keys := []string{usernameKey}
+		for k := range requiredValues {
+			keys = append(keys, k)
+		}
+		m, err = a.createDischargeMacaroon(ctx, requiredSourceModelUUID, requiredOffer, requiredRelation, userTag, keys, op, version)
+	} else {
+		var existingDeclaredCaveats []checkers.Caveat
+		for k, v := range declared {
+			existingDeclaredCaveats = append(existingDeclaredCaveats, checkers.DeclaredCaveat(k, v))
+		}
+		m, err = a.createDischargeMacaroonForExternalUser(ctx, requiredOffer, userTag, version, existingDeclaredCaveats...)
+	}
 	if err != nil {
-		return nil, errors.Trace(err)
+		err = errors.Annotate(err, "cannot create macaroon")
+		authlogger.Errorf("cannot create cross model macaroon: %v", err)
+		return nil, err
 	}
-	bakery, err := a.bakery.ExpireStorageAfter(localOfferPermissionExpiryTime)
+
+	return nil, &apiservererrors.DischargeRequiredError{
+		Cause:          cause,
+		Macaroon:       m,
+		LegacyMacaroon: m.M(),
+	}
+}
+
+func (a *authenticator) createDischargeMacaroon(
+	ctx context.Context, sourceModelUUID, offerUUID, relationKey string, userTag names.UserTag, requiredKeys []string, op bakery.Op, version bakery.Version,
+) (*bakery.Macaroon, error) {
+	authYaml, err := a.ctxt.offerPermissionYaml(sourceModelUUID, userTag.Id(), offerUUID, relationKey, permission.ConsumeAccess)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	keys := []string{usernameKey}
-	for k := range requiredValues {
-		keys = append(keys, k)
+	bakery, err := a.bakery.ExpireStorageAfter(offerPermissionExpiryTime)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	m, err := bakery.NewMacaroon(
+	return bakery.NewMacaroon(
 		ctx,
 		version,
 		[]checkers.Caveat{
 			checkers.NeedDeclaredCaveat(
 				checkers.Caveat{
-					Location:  a.offerAccessEndpoint,
+					Location:  a.localOfferAccessEndpoint,
```

Comment:
I think we can remove the `localOfferAccessEndpoint` attribute from the authenticator and instead just get it from the AuthContext. It seems we are storing the endpoint (for local) in 2 places and just using the one on the context makes the local case and the jaas case the same.

We want the authenticator itself to be common and ideally the same for local vs jaas, and the auth context has the differences.

ie for jaas
`Location:  a.ctxt.jaasOfferAccessEndpoint,`

for local

`Location:  a.ctxt.localOfferAccessEndpoint,`

So the logic for `(a *authenticator) createDischargeMacaroon` would just delegate to a method on the context. There would not be this logic `if a.ctxt.jaasOfferBakery == nil` in the authenticator at all.